### PR TITLE
Scaladoc: under `-release`, fix generation of links to JVM classes

### LINKED ISSUE #12820: Scaladoc generates incorrect link to JDK API docs with `-release` option
## Reproduction steps

Scala version: 2.13.11
JDK version: 17.0.7 (OpenJDK)
sbt version: 1.9.1

```scala
/** Example: [[java.util.concurrent.atomic.AtomicReference]]. */
class Example
```

`build.sbt`:
```scala
scalaVersion := "2.13.11"

scalacOptions ++= Seq("-release", "11")
```

## Problem

The generated API documentation (with `sbt doc`) contains an incorrect (dead) link to the `AtomicReference` apidoc. The incorrect link is https://docs.oracle.com/en/java/javase/17/docs/api/java/util/concurrent/atomic/AtomicReference.html (there is a `java.base` missing). Also note that the version is `17`, regardless of the `-release 11` setting.

_Without_ the `-release` setting, the generated link is correct: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/atomic/AtomicReference.html.

Originally I've opened an issue [here](https://github.com/typelevel/sbt-typelevel/issues/592), but it seems the problem is not sbt-plugin specific (probably not even sbt specific).

### PR DESCRIPTION
**Pull request by Math-ias**
Fixes scala/bug#12820.

### CODE REVIEW

**Comment by som-snytt** [comment]
ct.sym has module info.
```
00000010  00 00 00 00 00 00 00 00  00 00 23 00 00 00 38 2f  |..........#...8/|
00000020  6a 61 76 61 2e 61 63 74  69 76 61 74 69 6f 6e 2f  |java.activation/|
00000030  6a 61 76 61 78 2f 61 63  74 69 76 61 74 69 6f 6e  |javax/activation|
00000040  2f 03 00 50 4b 07 08 00  00 00 00 02 00 00 00 00  |/..PK...........|
```
oh apparently I dumped it out already a year ago Halloween in my `/tmp/ct` which I learned only after trying that path again. So I've forgotten what I tried before.

I see `K/java.base` for example and `K/java.base/module-info.sig`.

This is where it indexes the available packages:
https://github.com/scala/scala/blob/v2.13.15/src/compiler/scala/tools/nsc/classpath/DirectoryClassPath.scala#L285

It could also preserve modules. The modules needn't be exposed through the ClassPath interface; Scaladoc could just use CtSym directly.

I remember this line to switch class path representations, so it's not unprecedented to be ct.sym-aware:
https://github.com/scala/scala/blob/v2.13.15/src/compiler/scala/tools/nsc/Global.scala#L150

I'm dubious about using `Class.forName` here, as that uses the tool's class loader. It might kind of work.

The tool itself must be JDK 8 compatible, per the CI failure.

**Comment by Math-ias** [comment]
> ct.sym has module info.

I did not know the ct.sym file had module info, I will take a look at using it. I did find [the openjdk tool that makes that file for a reference](https://github.com/openjdk/jdk/blob/master/make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java).

Though parsing the `ct.sym` will crack only half the examples (better than none). I've come up with a different replication. https://github.com/scala/bug/issues/12820#issuecomment-2406023982

> I'm dubious about using Class.forName here, as that uses the tool's class loader. It might kind of work.

I tried to generalize my solution to also link external javadocs correctly. The external example class I tried couldn't be found on the tool's class loader but I didn't dig further. I didn't know the tool's class loader might have different access. I suppose it kind of worked for me in my limited tests because stdlib classes are generally present.

The java 8 requirement means I'll go back to the drawing board though. `ct.sym` or something else.

**Comment by SethTisue** [comment]
> The tool itself must be JDK 8 compatible, per the CI failure

Sigh... that would be a sad constraint for us to have live under more or less forever (since it's not clear when or if Scala 2 might drop Java 8 support).

If we accept that constraint, then there's no way for us to do anything at all with JPMS.

Perhaps @lrytz would like to weigh in on this. If behaving reasonably on JDK 9+ requires us to write some JDK 9+ specific code, then I would hope we could find some way to accommodate that code in this project. I would suggest proceeding with your fix on the assumption that we will find a way to incorporate the fix if we need to, even if it doesn't compile on Java 8.

**Comment by SethTisue** [comment]
> I did not know the ct.sym file had module info, I will take a look at using it

Hmm, so does that mean you're not currently stuck? I'm sorry I haven't offered help sooner (but I'm returning from a trifecta of illness, vacation, and sabbatical).

**Comment by lrytz** [comment]
> If behaving reasonably on JDK 9+ requires us to write some JDK 9+ specific code

Right, so if the compiler runs on Java 8 there's no ct.sym file. We can access the necessary API reflectively if running on Java 9+, it seems to be only `Class.getModule()` and `module.getName()`.

**Comment by SethTisue** [comment]
thanks, reflection is a good suggestion if the amount of code involved is small enough

**Comment by Math-ias** [comment]
> no way for us to do anything at all with JPMS.
> so does that mean you're not currently stuck?

Yes in the meantime I've been researching how to read [java module information](https://docs.oracle.com/javase/specs/jvms/se23/html/jvms-4.html#jvms-4.4.11) from class files. This is relevant because `ct.sym` seems to be _roughly_ a JAR (zip archive) of _roughly_ class files (sig files).

With my current notes and the latest discussion I think should ...
 * Save reading java modules from class files for a different issue
 * Narrow fix with if statements to fix _just_ the original reported issue (would make me feel better about risk).
 * Use reflection to _try_ to access java module methods.
 * Create a new issue to fix this in java 8 (maybe where my module reading goes).

**Comment by lrytz** [comment]
Is there anything to fix if scaladoc is running on Java 8? There are no modules there, so the API links are correct, no?

**Comment by Math-ias** [comment]
@lrytz good point!
```
ChoiceSetting("-release", "release", "Compile for a version of the Java API and target class file.", AllTargetVersions, normalizeTarget(javaSpecVersion))
      .withPostSetHook { setting =>
        val current = setting.value.toInt
        if (!isJavaAtLeast("9") && current > 8) errorFn.apply("-release is only supported on JVM 9 and higher")
```
I'm just about ready. All that's left is linking to the targeted JDK's javadocs.

**Comment by Math-ias** [comment]
I notice this "combined" step marked with an x saying a previous commit has failures. Does this mean I need to squash commits?

**Comment by SethTisue** [comment]
Before we hit "merge" every commit needs to individually pass CI in sequence. You could rebase now, or wait, either way. (Sometimes it's easier for reviewers if you hold off on rebasing until the end, if there's been review feedback and then new commits that specifically address that feedback. Or you might decide you'd rather just have reviewers look at the cleaned-up version.)

File: src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala [hunk 1]
```diff
@@ -60,9 +60,9 @@ trait MemberLookup extends base.MemberLookupBase {
         /* Get package object which has associatedFile ne null */
         sym.info.member(newTermName("package"))
       else sym
-    classpathEntryFor(sym1) flatMap { path =>
+    classpathEntryFor(sym1) flatMap { case path =>
```
Comment (SethTisue) [inline src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala]: I don't see a reason to add this?

File: src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala [hunk 2]
```diff
@@ -100,11 +100,36 @@ trait MemberLookup extends base.MemberLookupBase {
   private def isJDK(sym: Symbol) =
     sym.associatedFile.underlyingSource.map(f => isChildOf(f, (sys.props("java.home")))).getOrElse(false)
 
-  def jdkUrl(path: String): String = {
-    if (path.endsWith(".jmod")) {
+  // ISSUE-12820
```
Comment (SethTisue) [inline src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:103]: the new code beginning here is incorrectly indented, mind correcting that?

File: src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala (hunk 1 above)
Comment (Math-ias) [reply src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala]: Pushed to remove this. I was experimenting with if I could make the if statements clearer but decided against the change and forgot to remove all my changes.

File: src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala (hunk 2 above)
Comment (Math-ias) [reply src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:103]: Passed the new code into scalafmt like `pbpaste | scalafmt --stdin --stdout` to help me find and auto fix errors. Used the following `.scalafmt.conf`.
```
version = 3.8.4-RC1
runner.dialect = scala213
```

Comment (SethTisue) [reply src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:103]: I'm sorry, I was wrong

File: src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala [hunk 3]
```diff
@@ -100,11 +100,35 @@ trait MemberLookup extends base.MemberLookupBase {
...
+    val clazz = Class.forName("java.lang.Class")
+    clazz.getMethod("getModule")
+  }.toOption
+  private lazy val moduleGetName = Try {
+    val clazz = Class.forName("java.lang.Module")
+    clazz.getMethod("getName")
+  }.toOption
+
+  def jdkUrl(path: String, sym: Symbol): String = {
+    if (path.endsWith(".jmod") && javaVersion > 11) {
```
Comment (SethTisue) [inline src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala]: maybe I'm missing something obvious, but why `> 11` and not `>= 11` (or even `>= 9`)? Wasn't this stuff added in JDK 9?

I don't actually care about JDK 9 or 10, but I am wondering why 11 is excluded here.

Comment (Math-ias) [reply src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala]: Oh my goodness, very embarrassing for me, it should be `>= 11`. I sneak it into one of commit messages, but javadocs _appear_ to start using the module name as part of the path starting in 11 even though modules are added in 9. I can't find a formal announcement of this change, just something I noticed while testing. Though it appears I didn't test enough. Compare links below ...

```
https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/lang/Integer.html
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Integer.html
https://docs.oracle.com/javase/10/docs/api/java/lang/Integer.html
https://docs.oracle.com/javase/9/docs/api/java/lang/Integer.html
https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html
```

Comment (Math-ias) [reply src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala]: Changed the kind of equality used in the latest force push.

Comment (SethTisue) [reply src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala]: (pats self on back and says "good eye" 😁 )

File: src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala [hunk 4]
```diff
@@ -128,11 +152,15 @@ trait MemberLookup extends base.MemberLookupBase {
...
-      case _                       => 8   // shrug!
+    global.settings.releaseValue
+      .getOrElse(System.getProperty("java.specification.version"))
+      .split('.')
+      .take(2)
+      .map(_.toIntOption) match {
+      case Array(Some(1), Some(n)) => n // example: 1.8.0_242
+      case Array(Some(n))          => n // example: 14
+      case Array(Some(n), _)       => n // example: 11.0.7
+      case _                       => 8 // shrug!
```
Comment (som-snytt) [inline src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:163]: Should use `scala.util.Properties` instead
```
scala> util.Properties.javaSpecVersion
val res0: String = 23

scala> util.Properties.isJavaAtLeast(11)
val res1: Boolean = true
```

Comment (Math-ias) [reply src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:163]: I've pushed changes that replace `System.getProperty("java.specification.version")` with `scala.util.Properties.javaSpecVersion`. I couldn't replace version checks in this file e.g ...

```
if (javaVersion < 11) {
      s"https://docs.oracle.com/javase/$javaVersion/docs/api"
```

with `isJavaAtLeast` calls. `javaSpecVersion` seems to be pinned to the version of the JVM I'm running the scaladoc tool with, whereas a lot of these checks are for what version of JVM we're targeting with the scaladoc tool (specified with `-release`).

**Comment by Math-ias** [comment]
Thank you all for bearing with me and my mistakes in my first PR. 😄

**Comment by Math-ias** [comment]
Force pushed to slightly clean up the author section, I had two username/email combos.

**Comment by SethTisue** [comment]
Thank you! Thanks for sticking with this.

**Pull request merged**